zebra-network/src/peer_set/set.rs

```diff
@@ -441,14 +424,26 @@ where
     ///
     /// Move newly ready services to the ready list if they are for peers with supported protocol
     /// versions, otherwise they are dropped. Also drop failed services.
-    fn poll_unready(&mut self, cx: &mut Context<'_>) {
-        loop {
-            match Pin::new(&mut self.unready_services).poll_next(cx) {
-                // No unready service changes, or empty unready services
-                Poll::Pending | Poll::Ready(None) => return,
+    ///
+    /// Never returns an error. If there are no unready peers, returns `Ok`, and preserves the
+    /// original waker. Otherwise, returns `Poll::Pending`, and registers a wakeup for the next
+    /// peer that becomes ready.
+    fn poll_unready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), BoxError>> {
+        if self.unready_services.is_empty() {
+            // Don't replace the previous waker, because we have no peers to wake the task.
```

Comment:
This is unnecessary, because the waker is a copy, so it can be woken by the spawned task for unready peers, or the inventory timer:
https://github.com/ZcashFoundation/zebra/pull/7859/files#r1380722199